File: py/measurement_protocol_test.py
# Copyright 2024 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Unit tests for measurement_protocol.py."""

import requests_mock

from py import measurement_protocol
from google3.pyglib import resources
from absl.testing import absltest


class MeasurementProtocolTest(absltest.TestCase):

  def setUp(self):
    """Returns a Tadau instance.

    Returns:

    measurement_protocol.Tadau
    """
    super().setUp()
    self.tadau = measurement_protocol.Tadau(
        api_secret='1232', measurement_id='G-1223214', opt_in=True
    )

  def test_event_name_validation(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_events([{
          'client_id': '123',
          'name': 'event name 1',
          'value': '42',
          'important_event': 'False',
          'user_id': '11',
      }])

      self.assertEqual(
          m.last_request.json()['events'][0]['name'], 'eventname1'
      )

  def test_fail_safe_with_required_params(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_events([
          {
              'client_id': '123',
              'name': 'event_name',
              'value': '42',
              'important_event': 'False',
              'user_id': '11',
          },
          {
              'name': 'event_name',
              'successful': '42',
              'important_event': 'False',
          },
      ])

      self.assertEqual(m.call_count, 2)

  def test_successful_triggered_events(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_events([
          {
              'client_id': '123',
              'name': 'event_name',
              'value': '42',
              'important_event': 'False',
              'user_id': '11',
          },
          {
              'client_id': '4321',
              'name': 'event_name',
              'value': '42',
              'important_event': 'False',
          },
      ])

      self.assertEqual(m.call_count, 2)
      self.assertNotIn('user_id', m.last_request.json().keys())

  def test_incorrect_load(self):
    with self.assertRaises(AssertionError):
      measurement_protocol.Tadau(opt_in=True)

  def test_load_config_file(self):
    config_file = resources.GetResourceFilename(
        'google3/corp/gtech/ads/infrastructure/tadau/py/config_test.yaml'
    )
    tadau = measurement_protocol.Tadau(
        config_file_location=config_file
    )

    self.assertEqual(tadau.api_secret, '1232')
    self.assertEqual(tadau.measurement_id, 'G-1223214')
    self.assertEqual(tadau.opt_in, True)
    self.assertEqual(
        tadau.fixed_dimensions.get('deploy_id'),
        'bdb40a38-f845-4c04-abdc-1a51528d45e2',
    )

  def test_fixed_dimensions(self):

    tadau_with_fixed_dimensions = measurement_protocol.Tadau(
        api_secret='1232',
        measurement_id='G-1223214',
        opt_in=True,
        fixed_dimensions={'deploy_id': '123456asc'},
    )
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      tadau_with_fixed_dimensions.send_events([
          {
              'client_id': '123',
              'name': 'event name 1',
              'value': '42',
              'important_event': 'False',
              'user_id': '11',
          },
          {
              'client_id': '4321',
              'name': 'event name 2',
              'value': '42',
              'important_event': 'False',
          },
      ])

      request_history = m.request_history

      self.assertEqual(
          request_history[0].json()['events'][0]['params']['deploy_id'],
          '123456asc',
      )

      self.assertEqual(
          request_history[1].json()['events'][0]['params']['deploy_id'],
          '123456asc',
      )

  def test_ads_event(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_ads_event(
          'audience-created',
          'data integration',
          'GAds',
          '123456789',
          'audienceList',
          '9812176317',
      )

      self.assertEqual(m.last_request.json()['events'][0]['name'], 'ads-event')
      self.assertEqual(
          m.last_request.json()['events'][0]['params'][
              'event_is_impact_action'
          ],
          True,
      )

  def test_custom_event(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_custom_event('leads-generated', False, '')

      self.assertEqual(
          m.last_request.json()['events'][0]['name'], 'custom-event'
      )

  def test_error_event(self):
    with requests_mock.Mocker() as m:
      m.post(requests_mock.ANY, status_code=204)

      self.tadau.send_error_event(
          'conversion action not found', '404', '', 'conversionUploader.py'
      )

      self.assertEqual(
          m.last_request.json()['events'][0]['name'], 'error-event'
      )

  def test_not_opted_in_load(self):

    with self.assertRaises(AssertionError):
      tadau = measurement_protocol.Tadau()

      with requests_mock.Mocker() as m:
        m.post(requests_mock.ANY, status_code=204)

        tadau.send_ads_event(
            'audience-created',
            'data integration',
            'GAds',
            '123456789',
            'audienceList',
            '9812176317',
        )

      self.assertEqual(m.call_count, 0)

if __name__ == '__main__':
  absltest.main()

File: ts/measurement_protocol_test.ts
/**
 * @license
 * Copyright 2024 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {setUpAppsScriptSimulator} from 'google3/javascript/apps/maestro/simulator/closure_apps_script_simulator-closurized';
import 'jasmine';
import {Tadau} from './measurement_protocol';

describe('Tadau', () => {
  let tadau: Tadau;
  let urlFetchAppSpy: jasmine.Spy;

  beforeEach(() => {
    setUpAppsScriptSimulator();
    tadau = new Tadau({
      apiSecret: 'test_api_secret',
      measurementId: 'test_measurement_id',
      optIn: true,
    });
    spyOn(Utilities, 'getUuid').and.returnValue('test_client_id');
    urlFetchAppSpy = spyOn(UrlFetchApp, 'fetch').and.returnValue(
      createFakeResponse(200),
    );
  });
  function createFakeResponse(
    responseCode: number,
  ): GoogleAppsScript.URL_Fetch.HTTPResponse {
    return {
      getResponseCode: () => responseCode,
    } as GoogleAppsScript.URL_Fetch.HTTPResponse;
  }

  it('should be instantiated', () => {
    expect(tadau).toBeDefined();
  });

  it('should send events', () => {
    const events = [
      {
        'name': 'test-event',
        'test_param': 'test_value',
      },
    ];

    tadau.sendEvents(events);

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'test-event',
              'params': {
                'test_param': 'test_value',
              },
            },
          ],
        }),
      },
    );
  });

  it('should send ads event', () => {
    tadau.sendAdsEvent(
      'test_action',
      'test_context',
      'test_platform',
      'test_platform_id',
      'test_resource',
      'test_resource_id',
    );

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'ads-event',
              'params': {
                'event_is_impact_action': true,
                'event_action': 'test_action',
                'event_context': 'test_context',
                'ads_platform': 'test_platform',
                'ads_platform_id': 'test_platform_id',
                'ads_resource': 'test_resource',
                'ads_resource_id': 'test_resource_id',
              },
            },
          ],
        }),
      },
    );
  });

  it('should send custom event', () => {
    tadau.sendCustomEvent('test_action', true, 'test_context');

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'custom-event',
              'params': {
                'event_action': 'test_action',
                'event_is_impact_action': true,
                'event_context': 'test_context',
              },
            },
          ],
        }),
      },
    );
  });

  it('should send error event', () => {
    tadau.sendErrorEvent(
      'test_error_message',
      'test_error_code',
      'test_error_location',
      'test_error_location_id',
    );

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'error-event',
              'params': {
                'error_message': 'test_error_message',
                'error_code': 'test_error_code',
                'error_location': 'test_error_location',
                'error_location_id': 'test_error_location_id',
              },
            },
          ],
        }),
      },
    );
  });

  it('should not send events when not opt-in.', () => {
    const events = [
      {
        'name': 'test-event',
        'test_param': 'test_value',
      },
    ];
    tadau = new Tadau({
      apiSecret: 'test_api_secret',
      measurementId: 'test_measurement_id',
      optIn: false,
    });
    tadau.sendEvents(events);

    expect(urlFetchAppSpy).not.toHaveBeenCalled();
  });

  it('should throw error if api secret or measurement id is invalid', () => {
    expect(() => {
      tadau = new Tadau({
        apiSecret: '',
        measurementId: 'test_measurement_id',
      });
    }).toThrowError(
      'Tadau: Class could not be initiated because api_secret and/or measurement_id is invalid',
    );

    expect(() => {
      tadau = new Tadau({
        apiSecret: 'test_api_secret',
        measurementId: '',
      });
    }).toThrowError(
      'Tadau: Class could not be initiated because api_secret and/or measurement_id is invalid',
    );
  });

  it('should not send events if events is empty', () => {
    tadau.sendEvents([]);

    expect(urlFetchAppSpy).not.toHaveBeenCalled();
  });

  it('should not send events if event is invalid', () => {
    tadau.sendEvents([
      {
        'app_instance_id': 'reserverd-key',
      },
    ]);

    expect(urlFetchAppSpy).not.toHaveBeenCalled();
  });

  it('should send events with fixed dimensions', () => {
    tadau = new Tadau({
      apiSecret: 'test_api_secret',
      measurementId: 'test_measurement_id',
      fixedDimensions: {
        'test_fixed_dimension': 'test_fixed_value',
      },
      optIn: true,
    });

    const events = [
      {
        'name': 'test-event',
        'test_param': 'test_value',
      },
    ];

    tadau.sendEvents(events);

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'test-event',
              'params': {
                'test_fixed_dimension': 'test_fixed_value',
                'test_param': 'test_value',
              },
            },
          ],
        }),
      },
    );
  });

  it('should send events with fixed dimensions from properties', () => {
    const scriptProperties = PropertiesService.getScriptProperties();
    scriptProperties.setProperties({
      'apiSecret': 'test_api_secret',
      'measurementId': 'test_measurement_id',
      'deployId': 'fake_id',
      'deployInfra': 'fake_infra',
      'deployCreatedTime': 'fake_created_time',
      'deployUpdatedTime': 'fake_updated_time',
    });
    tadau = new Tadau({
      loadConfigFromScriptProperties: true,
      optIn: true,
    });

    const events = [
      {
        'name': 'test-event',
        'test_param': 'test_value',
      },
    ];

    tadau.sendEvents(events);

    expect(urlFetchAppSpy).toHaveBeenCalledWith(
      'https://www.google-analytics.com/mp/collect?api_secret=test_api_secret&measurement_id=test_measurement_id',
      {
        method: 'post',
        payload: JSON.stringify({
          'non_personalized_ads': true,
          'client_id': 'test_client_id',
          'events': [
            {
              'name': 'test-event',
              'params': {
                'deployId': 'fake_id',
                'deployInfra': 'fake_infra',
                'deployCreatedTime': 'fake_created_time',
                'deployUpdatedTime': 'fake_updated_time',
                'test_param': 'test_value',
              },
            },
          ],
        }),
      },
    );
  });
});
